examples/generative/diffusion/fid.py

```diff
@@ -0,0 +1,202 @@
+    # This is a direct PyTorch translation of http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz
+    logger0.info("Loading Inception-v3 model...")
+    detector_url = "https://api.ngc.nvidia.com/v2/models/nvidia/research/stylegan3/versions/1/files/metrics/inception-2015-12-05.pkl"
+    detector_kwargs = dict(return_features=True)
+    feature_dim = 2048
+    with open_url(detector_url, verbose=(dist.get_rank() == 0)) as f:
+        detector_net = pickle.load(f).to(device)
+
+    # List images.
+    logger0.info(f'Loading images from "{image_path}"...')
+    dataset_obj = ImageFolderDataset(
+        path=image_path, max_size=num_expected, random_seed=seed
+    )
+    if num_expected is not None and len(dataset_obj) < num_expected:
+        raise ValueError(
+            f"Found {len(dataset_obj)} images, but expected at least {num_expected}"
+        )
+    if len(dataset_obj) < 2:
+        raise ValueError(
+            f"Found {len(dataset_obj)} images, but need at least 2 to compute statistics"
+        )
+
+    # Other ranks follow.
+    if dist.world_size > 1 and dist.rank == 0:
+        torch.distributed.barrier()
+
+    # Divide images into batches.
+    num_batches = (
+        (len(dataset_obj) - 1) // (max_batch_size * dist.world_size) + 1
+    ) * dist.world_size
+    all_batches = torch.arange(len(dataset_obj)).tensor_split(num_batches)
+    rank_batches = all_batches[dist.rank :: dist.world_size]
+    data_loader = torch.utils.data.DataLoader(
+        dataset_obj,
+        batch_sampler=rank_batches,
+        num_workers=num_workers,
+        prefetch_factor=prefetch_factor,
+    )
+
+    # Accumulate statistics.
+    logger0.info(f"Calculating statistics for {len(dataset_obj)} images...")
+    mu = torch.zeros([feature_dim], dtype=torch.float64, device=device)
+    sigma = torch.zeros([feature_dim, feature_dim], dtype=torch.float64, device=device)
+    for images, _ in tqdm.tqdm(
+        data_loader, unit="batch", disable=(dist.get_rank() != 0)
+    ):
+        if dist.world_size > 1:
+            torch.distributed.barrier()
+        if images.shape[0] == 0:
+            continue
+        if images.shape[1] == 1:
+            images = images.repeat([1, 3, 1, 1])
+        features = detector_net(images.to(device), **detector_kwargs).to(torch.float64)
+        mu += features.sum(0)
+        sigma += features.T @ features
+
+    # Calculate grand totals.
+    if dist.world_size > 1:
+        torch.distributed.all_reduce(mu)
+        torch.distributed.all_reduce(sigma)
+    mu /= len(dataset_obj)
+    sigma -= mu.ger(mu) * len(dataset_obj)
+    sigma /= len(dataset_obj) - 1
+    return mu.cpu().numpy(), sigma.cpu().numpy()
+
+
+def calculate_fid_from_inception_stats(mu, sigma, mu_ref, sigma_ref):
+    m = np.square(mu - mu_ref).sum()
+    s, _ = scipy.linalg.sqrtm(np.dot(sigma, sigma_ref), disp=False)
+    fid = m + np.trace(sigma + sigma_ref - s * 2)
+    return float(np.real(fid))
```

Comment:
1. This should be moved to `modulus/metrics`.
2. scipy.ilnalg.sqrtm for this matrix is not stable generally.